### ISSUE TITLE
ProjectDiscontinuosCoefficient in nonconforming meshes

### ISSUE THREAD

**Issue by MC3JMF** [mfem/mfem#443]
Hello!. I'm new in the MFEM community. I am working with the elasticity equation using a structured mesh of Quadrilateral Elements. After solving the equation, I am interesting in refining the mesh around a prescribed isoline of the strain energy field. To obtain the isoline I need the strain energy projected at nodes. So I do the following:

       H1_FECollection *pfec=new H1_FECollection(order, dim);  
       FiniteElementSpace *pfespace=new FiniteElementSpace(pmesh, pfec, 1); // dimension 1   
       GridFunction strainenerg=(pfespace); 
       StrainEnergyCoefficient strainenerg_coeff(x, lambda_c, mu_c);
       strainenerg.ProjectDiscCoefficient(strainenerg_coeff, mfem::GridFunction::ARITHMETIC); 

This works well until I perform a nonconforming refinement of the mesh. In this case the ProjectDiscCoefficient does not give a continuous field anymore. 

My question is : Am I doing what i want right or there is something I have missed ?

Thank you in advance !

**Comment by jakubcerveny** [comment]
Hi @MC3JMF, to force a solution to be continuous on a nonconforming mesh (i.e., to make sure the hanging nodes have the correct interpolated values), you first need to restrict a solution vector to the "true" DOFs only and then prolong again to get the correct values in the slave DOFs:
```
Vector tmp(fespace->TrueVSize());
fespace->GetRestrictionMatrix()->Mult(x, tmp);
fespace->GetProlongationMatrix()->Mult(tmp, x);
```
The first Mult throws away the slave DOFs and the second Mult does the interpolation. `x` is a `ParGridFunction` (a `Vector`, really), as in `ex1p`.

Hope that helps,
Jakub

**Comment by MC3JMF** [comment]
Thank you Jakub!, that solved the problem.

**Comment by stefanozampini** [comment]
I'm reopening the issue since I think this needs to be fixed.

Considering a sequential run, ProjectDiscCoefficient gives two different answers when called on a GridFunction or on a ParGridFunction. This is because ParallelAssembled is called for the ParGridFunction case https://github.com/mfem/mfem/blob/master/fem/pgridfunc.cpp#L357; in the GridFunction case, the slave dofs are left unmodified (as it should be IMHO) https://github.com/mfem/mfem/blob/master/fem/gridfunc.cpp#L1739

**Comment by stefanozampini** [comment]
@v-dobrev @tzanio what do you think?

**Comment by v-dobrev** [comment]
Hi @stefanozampini,

I think you are right that there is a bug in `ParGridFunction::ProjectDiscCoefficient`, if the method is used for a non-conforming mesh. Instead of calling `ParallelAssemble` we should use, instead, the `GroupCommunicator` to do the summation across processors. Until recently, the `GroupCommunicator` was undefined for non-conforming meshes - this method (`ProjectDiscCoefficient`) was written before that and (at the time) was only meant to work on conforming meshes. We can now fix this issue and have the same behavior in parallel as in serial.

Basically, we can replace https://github.com/mfem/mfem/blob/da7b500db5fd7cbdec586dcdb8f2b6fecbc56565/fem/pgridfunc.cpp#L371-L373
with the following:
```c++
   gcomm.Reduce(data, GroupCommunicator::Sum);
   gcomm.Bcast(data);
```
Do you want to try this fix and make a branch/PR?

**Comment by stefanozampini** [comment]
@v-dobrev Here you have 
https://github.com/mfem/mfem/commit/0f64f3e00680ee32a93626cb8be4798d1b56d3a5

The commit is part of a branch for which I plan to open a PR soon

**Comment by tzanio** [comment]
Addressed in #756
